service/matching/matching_engine_test.go

```diff
@@ -2761,6 +2762,321 @@ func (s *matchingEngineSuite) TestUpdatePhysicalTaskQueueGauge_BuildID() {
+	// increase and remains equal to config.RangeSize.
+	for i := int64(0); i < taskCount-1; i++ {
+		taskID := i + s.matchingEngine.config.RangeSize + 2
+		tlMgr.backlogMgr.taskAckManager.completeTask(taskID)
+	}
+
+	// Since the first task was not completed, the AckLevel did not go up leading to backlog size not decreasing
+	s.EqualValues(tlMgr.backlogMgr.db.getApproximateBacklogCount(), int64(taskCount))
+
+	// Adding a new task which should persist the backlog count and the ackLevel
+	s.addWorkflowTasks(false, 1, 1, taskQueue, workflowExecution, nil)
+
+	// unloading the TQM which also stops the backlogManager; set backlogManager.skipFinalUpdate to true to not persist
+	// the latest ackLevel
+	tlMgr.backlogMgr.skipFinalUpdate.Store(true)
+	tqm, err := s.matchingEngine.getTaskQueuePartitionManager(context.Background(), ptq.Partition(), false)
+	s.NoError(err)
+	s.matchingEngine.unloadTaskQueuePartition(tqm)
+
+	// Loading the TQM for the same TQ partition
+	tqm_loaded, err := s.matchingEngine.getTaskQueuePartitionManager(context.Background(), ptq.Partition(), true)
+	s.NoError(err)
+	tqm_loaded_Impl, ok := tqm_loaded.(*taskQueuePartitionManagerImpl)
+	s.Equal(ok, true)
+
+	tlMgr, ok = tqm_loaded_Impl.defaultQueue.(*physicalTaskQueueManagerImpl)
+	s.Equal(ok, true)
+	s.EqualValues(tlMgr.backlogMgr.db.getApproximateBacklogCount(), taskCount+1) // should read "taskCount + 1", not 1
+
+}
+
+// Concurrent tests for testing approximateBacklogCounter
+
+// TestConcurrentAddWorkflowTasksValidateBacklogCounter adds multiple workflow tasks concurrently to
+// persistence and checks if the backlog counter was updated accordingly.
+func (s *matchingEngineSuite) TestConcurrentAddWorkflowTasksValidateBacklogCounter() {
+	_, workflowExecution := s.generateWorkflowExecution()
+	taskQueue, ptq := s.createTQAndPTQForBacklogTests()
+
+	const workerCount = 20
+	const taskCount = 100
+
+	var wg sync.WaitGroup
+	wg.Add(workerCount)
+
+	s.addWorkflowTasks(true, workerCount, taskCount, taskQueue, workflowExecution, &wg)
+	wg.Wait()
+
+	// validating the approximateBacklogCounter
+	tlMgr := s.getPhysicalTaskQueueManagerImpl(ptq)
+	s.EqualValues(tlMgr.backlogMgr.db.getApproximateBacklogCount(), int64(taskCount*workerCount))
+}
+
+func (s *matchingEngineSuite) TestConcurrentAddCompleteWorkflowTasksValidateBacklogCounter() {
+	workflowType, workflowExecution := s.generateWorkflowExecution()
+	taskQueue, ptq := s.createTQAndPTQForBacklogTests()
+
+	const workerCount = 20
+	const taskCount = 100
+
+	var wg sync.WaitGroup
+	wg.Add(2 * workerCount)
+
+	s.addWorkflowTasks(true, workerCount, taskCount, taskQueue, workflowExecution, &wg)
+	s.mockHistoryWhilePolling(workflowType)
+	s.pollWorkflowTasks(true, workerCount, taskCount, nil, taskQueue, &wg)
+
+	wg.Wait()
+	s.EqualValues(0, s.taskManager.getTaskCount(ptq))
+	totalTasks := taskCount * workerCount
+	persisted := s.taskManager.getCreateTaskCount(ptq)
+	s.True(persisted < totalTasks)
```

Comment:
No, the purpose of this log was to check if not all tasks were sync matched and some were indeed spooled to the database so that they can be fetched later on.